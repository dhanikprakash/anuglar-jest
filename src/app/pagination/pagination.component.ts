import { CommonModule } from '@angular/common';
import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { RouterLink } from '@angular/router';
import { UtilsService } from '../utils.service';

@Component({
  selector: 'app-pagination',
  templateUrl: './pagination.component.html',
  styleUrls: ['./pagination.component.css']
})
export class PaginationComponent implements OnInit {

  @Input() total: number = 20;
  @Input() limit: number = 10;
  @Input() currentPage: number = 1;

  @Output('pageChange')
  pageChangeEvent = new EventEmitter<number>();

  pagesCount: number = 1;
  pages: number[] = [];

  constructor(private utilsService: UtilsService) {}

  ngOnInit(): void {
    this.pagesCount = Math.ceil(this.total / this.limit);
    this.pages =
      this.pagesCount > 0
        ? this.utilsService.range(1, this.pagesCount + 1)
        : [];
  }

  selectPage(page: number): void {
    this.pageChangeEvent.emit(page);
  }
}