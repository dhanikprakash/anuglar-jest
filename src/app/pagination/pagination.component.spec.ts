import { ComponentFixture, TestBed } from '@angular/core/testing';

import { PaginationComponent } from './pagination.component';
import { By } from '@angular/platform-browser';
import { first } from 'rxjs';
import { UtilsService } from '../utils.service';

describe('PaginationComponent', () => {
  let component: PaginationComponent;
  let fixture: ComponentFixture<PaginationComponent>;
  const mockUtilsService = {
    range: () => [1,2,3,4,5],
  };

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [PaginationComponent],
      providers: [{ provide: UtilsService, useValue: mockUtilsService }],
    });
    fixture = TestBed.createComponent(PaginationComponent);
    component = fixture.componentInstance;
    component.total = 50;
    component.limit = 10;
    component.currentPage = 1;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('renders correct pagination', () => {
    const pageContainers = fixture.debugElement.queryAll(
      By.css('[data-testid="page-container"]')
    );

    expect(pageContainers.length).toBe(5);
    expect(pageContainers.at(0)?.nativeElement.textContent).toEqual(' 1 ');
  });

  it('should emit a clicked page', () => {
    const pageContainers = fixture.debugElement.queryAll(
      By.css('[data-testid="page-container"]')
    );
    let clickedPage: number | undefined;

    component.pageChangeEvent.pipe(first()).subscribe((page) => {
      clickedPage = page;
    });

    pageContainers.at(0)?.triggerEventHandler('click');

    expect(clickedPage).toEqual(1);
  });
});
