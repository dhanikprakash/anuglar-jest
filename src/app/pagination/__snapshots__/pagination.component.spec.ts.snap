// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationComponent should match snapshot 1`] = `
PaginationComponent {
  "__ngContext__": 7,
  "currentPage": 1,
  "limit": 10,
  "pageChangeEvent": EventEmitter_ {
    "__isAsync": false,
    "closed": false,
    "currentObservers": null,
    "hasError": false,
    "isStopped": false,
    "observers": [],
    "thrownError": null,
  },
  "pages": [
    1,
    2,
    3,
    4,
    5,
  ],
  "pagesCount": 5,
  "total": 50,
  "utilsService": {
    "range": [Function],
  },
}
`;
